near-sdk/src/types/gas.rs

```diff
@@ -0,0 +1,115 @@
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+
+/// Represents the amount of NEAR tokens in "gas units" which are used to fund transactions.
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    BorshSerialize,
+    BorshDeserialize,
+    Hash,
+    BorshSchema,
+)]
+pub struct Gas(u64);
+
+impl Gas {
+    pub const fn new(amount: u64) -> Self {
+        Self(amount)
+    }
+}
+
+impl Serialize for Gas {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.0.to_string())
```

Comment:
I wonder if we could use [smallstr](https://crates.io/crates/smallstr) for these kinds of fixed-size strings to avoid the `to_string` allocation underneath, but might not be necessary

Reply:
We don't even need smallstr here, the length is bounded (not sure what the actual bound is), so we can just use stack-allocated `[u8]`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=132595fc106b94493a5a240512ec9063

Reply:
I am conflicted about this  impl via `to_string`. I think it's the best we can do, but we must be aware of the drawbacks.

The problem we want to solve here is that we want to serialize this as string *in json*. The current impl uses strings for every serde format. So if someone wants to use, eg, `bincode` for their contract, they'll be surprised to see that gas is encoded as string into the binary format. 

I feel that https://github.com/near/near-sdk-rs/issues/408 might be the "proper" soulution here, but it's clear that we aren't going to do that short-medium term. 

A possible hack here is using https://docs.serde.rs/serde/trait.Serializer.html#method.is_human_readable as a proxy for "is this JSON", but I am not sure if that's actually better. 

So, yeah, nothing to improve here, but maybe a good place for "// NOTA BENE: " comment.

Reply:
Yeah, I think we can just ideate what is the best solution wrt #408 after this next release, as swapping the integer serialization at the serialization layer would also add some unintended behaviour that users may not expect and may want to eject from, and this release already has a lot of breaking changes.